### PR TITLE
Hamming loss and distance scripts

### PR DESCRIPTION
**Pull request by SSaishruthi**
Scripts for hamming loss and distance 
Colab: https://colab.research.google.com/drive/1Msuv5xUu7lu5wDH1ei-VOPB-UnBolDfB

### CODE REVIEW

**Comment by SSaishruthi** [comment]
@Squadrick 
I did not complete the scripts. Started a PR to track.
In issue, you have specified to use `tf.keras.metrics.Metric`. Do you mean to use like this?

```
class HammingLoss(tf.keras.metrics.Metric):
    def __init__(self, name='hamming_loss', dtype=None, mode='multiclass'):
```

If so, can we use `MeanMetricWrapper` class from addons.utils? 
I did not understand the comment.

**Comment by Squadrick** [comment]
@SSaishruthi Extend `MeanMetricWrapper`, sorry I was unclear.

File: tensorflow_addons/metrics/hamming.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import tensorflow as tf
+from tensorflow.keras.metrics import Metric
+from tensorflow_addons.metrics.utils import MeanMetricWrapper
+
+def hamming_distance(actuals, predictions):
```
Comment (Squadrick) [inline tensorflow_addons/metrics/hamming.py]: Make the params for `hamming_distance` and `hamming_loss` the same. Use `y_true`, `y_pred` or `actuals`, `predictions.`

File: tensorflow_addons/metrics/hamming.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+        print(nonzero)
+        return 1.0 - nonzero
+
+    else:
+        nonzero = tf.cast(tf.math.count_nonzero(y_true - y_pred, axis=-1),
+            tf.float32)
+        return nonzero / y_true.get_shape()[-1]
+
+
+class HammingLoss(tf.keras.metrics.Metric):
```
Comment (Squadrick) [inline tensorflow_addons/metrics/hamming.py]: Make this `MeanMetricWrapper`.

**Comment by SSaishruthi** [comment]
Thanks for the comments. Will work on this and update

File: tensorflow_addons/metrics/hamming.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import tensorflow as tf
+from tensorflow.keras.metrics import Metric
+from tensorflow_addons.metrics.utils import MeanMetricWrapper
+
+def hamming_distance(actuals, predictions):
+  result=tf.not_equal(actuals,predictions)
```
Comment (feihugis) [inline tensorflow_addons/metrics/hamming.py]: missing a blank space here `(actuals,predictions)`

File: tensorflow_addons/metrics/hamming.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+  not_eq = tf.reduce_sum(tf.cast(result, tf.float32))
+  ham_distance = tf.math.divide_no_nan(not_eq, len(result))
+  return ham_distance
+
+def hamming_loss(y_true, y_pred, mode='multiclass'):
```
Comment (feihugis) [inline tensorflow_addons/metrics/hamming.py]: Could you add the method docstring?

File: tensorflow_addons/metrics/hamming.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    if mode not in ['multiclass', 'multilabel']:
+        raise TypeError('mode must be: [multiclass, multilabel])')
+
+    if mode == 'multiclass':
+        nonzero = tf.cast(tf.math.count_nonzero(y_true * y_pred, axis=-1), tf.float32)
```
Comment (feihugis) [inline tensorflow_addons/metrics/hamming.py]: Is this line beyond the line length limitation (80 chars)?

File: tensorflow_addons/metrics/hamming.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+        print(nonzero)
```
Comment (feihugis) [inline tensorflow_addons/metrics/hamming.py]: What's print for?

File: tensorflow_addons/metrics/hamming.py (hunk 2 above)
Comment (feihugis) [reply tensorflow_addons/metrics/hamming.py]: Could you add the class docstring?

File: tensorflow_addons/metrics/hamming.py [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+        return 1.0 - nonzero
+
+    else:
+        nonzero = tf.cast(tf.math.count_nonzero(y_true - y_pred, axis=-1),
+            tf.float32)
```
Comment (feihugis) [inline tensorflow_addons/metrics/hamming.py]: The indentation here seems not right.

File: tensorflow_addons/metrics/hamming.py [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+
+def hamming_loss(y_true, y_pred, mode='multiclass'):
+    if mode not in ['multiclass', 'multilabel']:
+        raise TypeError('mode must be: [multiclass, multilabel])')
+
+    if mode == 'multiclass':
+        nonzero = tf.cast(tf.math.count_nonzero(y_true * y_pred, axis=-1), tf.float32)
+        print(nonzero)
+        return 1.0 - nonzero
+
```
Comment (feihugis) [inline tensorflow_addons/metrics/hamming.py]: May not need this blank line.

**Comment by SSaishruthi** [comment]
@feihugis Thanks for the review.

This is a placeholder PR and we were working separately to finalize the approach. This is not the final version. I am working on adding test cases, docstring and code cleanup.

**Comment by feihugis** [comment]
@SSaishruthi Oh, sorry that I did not notice `[WIP]`. Please ignore my comments and continue your work.

**Comment by SSaishruthi** [comment]
@feihugis That's totally fine. I did help in enhancing the PR. Thanks for the pointers.

**Comment by SSaishruthi** [comment]
@Squadrick 
Added test scripts. If this is ok, I will go ahead with final code cleanup and other file updates.

**Comment by Squadrick** [comment]
@SSaishruthi Can you resolve the branch conflicts?

**Comment by SSaishruthi** [comment]
@Squadrick Done.
If the test approach is ok, I will do the final code clean and add docstrings.

File: tensorflow_addons/metrics/hamming.py (hunk 1 above)
Comment (Squadrick) [reply tensorflow_addons/metrics/hamming.py]: @SSaishruthi update this

File: tensorflow_addons/metrics/hamming.py [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+def hamming_distance(actuals, predictions):
+    result = tf.not_equal(actuals, predictions)
+    not_eq = tf.reduce_sum(tf.cast(result, tf.float32))
+    ham_distance = tf.math.divide_no_nan(not_eq, len(result))
+    return ham_distance
+
+
+def hamming_loss_fn(y_true, y_pred, mode):
+    if mode not in ['multiclass', 'multilabel']:
+        raise TypeError('mode must be: [multiclass, multilabel])')
```
Comment (Squadrick) [inline tensorflow_addons/metrics/hamming.py]: Extra `)`

**Comment by SSaishruthi** [comment]
@Squadrick Updated docstring and readme

**Comment by SSaishruthi** [comment]
@Squadrick Is this good to go now?

**Comment by SSaishruthi** [comment]
@Squadrick Is this PR in good shape?

**Comment by Squadrick** [comment]
Add a test to check that the metric works with Keras Model API. It should be good to go after that.

**Comment by SSaishruthi** [comment]
@Squadrick Do you have an example that I can refer for testing with Keras model API? I did not understand the requirement completely

**Comment by SSaishruthi** [comment]
@Squadrick Added keras test. Can you please check this?

**Comment by SSaishruthi** [comment]
@seanpmorgan Can you please trigger the test?

**Comment by SSaishruthi** [comment]
Thanks @seanpmorgan 
Looks like the error is not just with this metric. Is there any change that I can/ need to do?

**Comment by SSaishruthi** [comment]
@Squadrick Looks like tests have passed. Is this good to merge this?

File: tensorflow_addons/metrics/hamming.py [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+
+
+def hamming_distance(y_true, y_pred):
+    """Computes hamming distance.
+
+    Hamming distance is for comparing two binary strings.
+    It is the number of bit positions in which two bits
+    are different.
+
+    :param y_true: actual target value
```
Comment (Squadrick) [inline tensorflow_addons/metrics/hamming.py]: Could you change the docstring to standardize it according to like `f_scores.py`? We need to change the other metrics to conform to this format as well.

File: tensorflow_addons/metrics/__init__.py [hunk 11]
```diff
@@ -21,4 +21,6 @@
 from tensorflow_addons.metrics.cohens_kappa import CohenKappa
 from tensorflow_addons.metrics.f_scores import F1Score, FBetaScore
 from tensorflow_addons.metrics.r_square import RSquare
+from tensorflow_addons.metrics.utils import MeanMetricWrapper
+from tensorflow_addons.metrics.hamming import HammingLoss, hamming_distance
```
Comment (Squadrick) [inline tensorflow_addons/metrics/__init__.py]: Could you arrange this in alphabetical order?

File: tensorflow_addons/metrics/README.md [hunk 12]
```diff
@@ -7,6 +7,7 @@
 |  f_scores| Saishruthi Swaminathan | saishruthi.tn@gmail.com|
 |  r_square| Saishruthi Swaminathan| saishruthi.tn@gmail.com|
 |  multilabel_confusion_matrix | Saishruthi Swaminathan | saishruthi.tn@gmail.com|
+|  Hamming metrics | Saishruthi Swaminathan | saishruthi.tn@gmail.com|
```
Comment (Squadrick) [inline tensorflow_addons/metrics/README.md]: ditto

Comment (Squadrick) [reply tensorflow_addons/metrics/README.md]: Rename `Hamming metrics` -> `hamming`

File: tensorflow_addons/metrics/hamming_test.py [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+        predictions = tf.constant([1, 0, 0, 0, 1, 0, 0, 1, 0, 1],
+                                  dtype=tf.int32)
+        test_result = hamming_distance(actuals, predictions)
+        self.assertAllClose(0.3, test_result, atol=1e-5)
+
+    # Keras model check
+    def test_keras_model(self):
+        model = tf.keras.Sequential()
+        model.add(layers.Dense(64, activation='relu'))
+        model.add(layers.Dense(64, activation='relu'))
```
Comment (Squadrick) [inline tensorflow_addons/metrics/hamming_test.py]: Delete this layer.

**Comment by SSaishruthi** [comment]
@Squadrick Can you please check if updates are ok?

**Comment by SSaishruthi** [comment]
@seanpmorgan @Squadrick Is anyone looking at the metrics PRs ?

File: tensorflow_addons/metrics/README.md [hunk 14]
```diff
@@ -5,16 +5,18 @@
...
 |  multilabel_confusion_matrix | Saishruthi Swaminathan | saishruthi.tn@gmail.com|
+|  r_square| Saishruthi Swaminathan| saishruthi.tn@gmail.com|
 
 ## Contents
 | Submodule | Metric  | Reference                               |
 |:----------------------- |:-------------------|:---------------|
 | cohens_kappa| CohenKappa|[Cohen's Kappa](https://en.wikipedia.org/wiki/Cohen%27s_kappa)|
 | f1_scores| F1 micro, macro and weighted| [F1 Score](https://en.wikipedia.org/wiki/F1_score)|
-| r_square| RSquare|[R-Sqaure](https://en.wikipedia.org/wiki/Coefficient_of_determination)|
+| hamming | Hamming loss and distance | [Hamming distance](https://en.wikipedia.org/wiki/Hamming_distance)|
```
Comment (seanpmorgan) [inline tensorflow_addons/metrics/README.md]: nit: Match the names `HammingLoss` and `hamming_distance`

File: tensorflow_addons/metrics/hamming.py [hunk 15]
```diff
@@ -0,0 +1,142 @@
...
+    predictions = tf.constant([[0.82, 0.5, 0.90, 0],
+                               [0, 1, 0.4, 0.98],
+                               [0.89, 0.79, 0, 0.3]],
+                               dtype=tf.float32)
+    hl.update_state(actuals, predictions)
+    print('Hamming loss: ', hl.result().numpy()) # 0.16666667
+    ```
+    """
+    if mode not in ['multiclass', 'multilabel']:
+        raise TypeError('mode must be: [multiclass, multilabel]')
```
Comment (seanpmorgan) [inline tensorflow_addons/metrics/hamming.py]: nit: mode must be: `multiclass` or `multilabel`. The current way looks as though it takes in a list

**Comment by SSaishruthi** [comment]
@seanpmorgan Sorry for the delay.
Updated readme and addressed conflicts

**Comment by SSaishruthi** [comment]
Is this good to merge?
